blocks/api/utils.js

```diff
@@ -60,16 +77,22 @@ export function isUnmodifiedDefaultBlock( block ) {
  *                                          as the icon for the block in the
  *                                          inserter, or element or an object describing the icon.
  *
- * @return {Object} Object describing the icon.
+ * @return {?Object} Object describing the icon or null if the icon is invalid and impossible to normalize.
  */
 export function normalizeIconObject( icon ) {
 	if ( ! icon ) {
 		return { src: 'block-default' };
 	}
-	if ( isString( icon ) || isFunction( icon ) || icon instanceof Component ) {
+	if (
+		isValidIcon( icon )
+	) {
 		return { src: icon };
 	}
 
+	if ( ! isValidIcon( icon.src ) ) {
+		return null;
```

Comment:
This one is quite unexpected in my opinion. Shouldn't we return a default icon instead and warn?

Reply:
Shouldn't we split validation and normalization in that case? You would use it as follows inside blocks API:
```js
if ( ! isValidIcon( settings.icon ) {
	console.error(
		'The icon passed is invalid.'
	);
	return;
}
settings.icon = normalizeIconObject( settings.icon );
```

Reply:
Yes separating the concerns is a good idea. Validation and normalization were separated. We validate after normalization so even if the icon is in the format icon: { background: ..., src: ... } and the src is invalid we catch the problem.